src/main/java/org/elasticsearch/cluster/metadata/MetaData.java

```diff
@@ -1393,6 +1386,47 @@ public Builder customs(ImmutableOpenMap<String, Custom> customs) {
             this.customs.putAll(customs);
             return this;
         }
+        /*@
+
+         */
+        public static Settings getUpdatedSettings(Settings settings, Settings existedSettings){
+            Set<String> removed =  Sets.newHashSet();
+            ImmutableSettings.Builder updatedSettingsBuilder = ImmutableSettings.settingsBuilder();
+            updatedSettingsBuilder.put(settings);
+            ImmutableSettings.Builder cleanedSourceBuilder = ImmutableSettings.settingsBuilder();
+            for(Map.Entry<String, String> setting : updatedSettingsBuilder.internalMap().entrySet()){
+                if(setting.getValue().equals(REMOVE_MARKER)){
+                    removed.add(setting.getKey());
+                }else{
+                    String key = setting.getKey();
+                    String value = setting.getValue();
+                    cleanedSourceBuilder.put(key, value);
+                    //avoid strange behavior of Java binder that can choose Object... notation of ctor
+                }
+            }
+            final Settings cleanedSourceSettings = cleanedSourceBuilder.build();
+            ImmutableSettings.Builder sourceBuilder = ImmutableSettings.settingsBuilder();
+            sourceBuilder.put(existedSettings);
+            ImmutableSettings.Builder cleanedBuilder = ImmutableSettings.settingsBuilder();
+            for(Map.Entry<String, String> setting : sourceBuilder.internalMap().entrySet()){
+                boolean include = true;
+                for(String prefix : removed){
+                    if(setting.getKey().startsWith(prefix)){
+                        include =false;
+                        break;
+                    }
+                }
+                if(include){
+                    String key = setting.getKey();
+                    String value = setting.getValue();
+                    cleanedBuilder.put(key, value);
+                    //avoid strange behavior of Java binder that can choose Object... notation of ctor
+                }
+            }
+            final Settings cleanedTargetSettings = cleanedBuilder.build();
+            final Settings updatedSettings = settingsBuilder().put(cleanedTargetSettings).put(cleanedSourceSettings).build();
+            return updatedSettings;
+        }
```

Comment:
It's better to isolate setting merging process to method - that gives as far more control on it. Here we realize support for __REMOVE__ logic but we can apply any other merging logic here further.

May be it's better to move it to some SettingUtuls but while for now it is used only in MetaData.Builder i keep it there